core/src/main/java/org/apache/calcite/sql/validate/DelegatingScope.java

```diff
@@ -214,6 +216,47 @@ public SqlValidator getValidator() {
     return validator;
   }
 
+  /**
+   *
+   * @param map A map from table names to {@code ScopeChild} objects, which
+   *            represents all possible tables which a particular column
+   *            could come from.
+   * @return    A new table name map with unknown tables filtered out
+   *            (unless only unknown tables are present, in which case it
+   *            leaves an arbitrary unknown table which is assumed to contain
+   *            the column we are trying to resolve).
+   *            This ensures that columns are not matched to unknown tables if
+   *            they can instead be matched to tables with known schemas.
+   */
+  public Map<String, ScopeChild> filterTableNames(Map<String, ScopeChild> map) {
+    Map<String, ScopeChild> filteredMap = new HashMap<String, ScopeChild>();
+    for (String key : map.keySet()) {
+      if (!containsUnknownNamespace(map.get(key))) {
+        filteredMap.put(key, map.get(key));
+      }
+    }
+    // If original map consists only of unknown tables,
+    // arbitrarily assume the column corresponds to the first unknown table.
+    // We cannot do any better than this without prior schema information.
+    if (filteredMap.size() == 0 && map.size() > 0) {
+      String arbitraryKey = map.keySet().iterator().next();
+      filteredMap.put(arbitraryKey, map.get(arbitraryKey));
+    }
+    return filteredMap;
+  }
+
+  /**
+   * Determines whether the given {@code ScopeChild} corresponds to the
+   * namespace of an unknown table.
+   */
+  public boolean containsUnknownNamespace(ScopeChild child) {
+    if (child.namespace != null) {
```

Comment:
nit: Negate this if-check and return early. Then you can unnest the logic within it.